### ISSUE TITLE
TransferUtility.UploadAsync still uses Content-Length

### ISSUE THREAD

**Issue by eocron** [aws/aws-sdk-net#1095]
For large streams it is incorrect to:
 - Buffer it entirely to file, when only transit needed. Not every system has resources for this.
 - Expect that it has content-length in general: zip stream, encryption stream, etc - doesn't have one.
 - Expect that it has any other method other than Read/Write/Close/Dispose (dependency for implementation is bad, and it backfired in this issue)

## Expected Behavior
TransferUtility.UploadAsync do not use Content-Length in ANY way. Only for optimizations, but not as limitation of functionality - no strong dependency on it.

## Current Behavior
Conten-Length is still used to calculate number of parts (why is it even needed? parts can also be streamed! ).
Uploading of stream with unknown length is **impossible** even with using of "parts".
If you pass -1 in content-length, it will not enqueue a single _UploadPartRequest_ so upload will not happen - MalformedXML error will return.

## Possible Solution
Remove all content-length checks, Remove while loop in _MultipartUploadCommand.ExecuteAsync_ and instead look for return value in Read method in Stream class, fill buffer until part is formed, then enqueue UploadPartRequest command for send.
```
var read = ReadAsync(buffer, ...)
if(read < buffer.Length)
{ 
    _queue.Enqueue(last UploadPartRequest)
}
else
{
    _queue.Enqueue(UploadPartRequest)
}

```
## Steps to Reproduce (for bugs)
                var id = Guid.NewGuid().ToString("N");
                using (var client = new AmazonS3Client(...))
                {
                    var tu = new TransferUtility(client, new TransferUtilityConfig()
                    {
                        MinSizeBeforePartUpload = -1 //required to skip content-length check, otherwise it will complain
                    });
                    var tuur = new TransferUtilityUploadRequest()
                    {
                        StorageClass = S3StorageClass.StandardInfrequentAccess,
                        AutoResetStreamPosition = false,
                        BucketName = _cfg.Bucket,
                        InputStream = stream, //stream without content-length support
                        CannedACL = S3CannedACL.PublicRead,
                        Key = id,
                    };
                    await tu.UploadAsync(tuur);
                    return id;
                }

Only solution I see now for myself - fork your repo and fix it for myself.

## Context
Upload a streamed data, as simple as that. 

## Your Environment
* AWSSDK.Core version used: 3.3.24.4
* Targeted .NET platform: .NET Core 2.1

**Comment by eocron** [comment]
Question is about TransferUtility. Particulary MultipartUploadCommand and **its implementation, not interface**. How is this backward compatability? You expecting someone put code behind used properties (content-length)?

**Comment by zhengchun** [comment]
I have encounter same problem. using `multipart/form-data` method to upload file, we don't know about uploaded file length, I don't want to save entire file into memory or save to local disk and then post to AWS S3 . when call a `TransferUtility.Upload` always got error. `: 'Unable to read data from the transport connection:

**Comment by zhengchun** [comment]
A few hour later, I found an alternative solution,[Upload an Object to an S3 Bucket Using a Presigned URL (AWS SDK for .NET)](https://docs.aws.amazon.com/AmazonS3/latest/dev/UploadObjectPreSignedURLDotNetSDK.html).It's looks don't need specific a `Content-Length` , that can solved streaming upload file problem.

```c#
 var url = client.GetPreSignedURL(request);
 var httpRequest = new HttpRequestMessage(HttpMethod.Put, url);
httpRequest.Content = new StreamContent(stream);
var httpClient = new HttpClient();
var response = httpClient.SendAsync(httpRequest).GetAwaiter().GetResult();
```
Hope this can help you😄

**Comment by swASO** [comment]
same problem here...

**Comment by repspark-jd** [comment]
Also having the above issue; we are trying to not save the file into memory/local disk.

**Comment by zhengchun** [comment]
Hello everyone, I solved my problem a month ago, It works on [Bingdesktop.com](https://bingdesktop.com) that uploading a lot of large images file without causing memory leak. My front-end using `dropzonejs` library that using `multipart/form-data` method.  I am shared main code on gist. if anyone interesting,you can checking here: https://gist.github.com/zhengchun/1ae317ee3fb623da4ae3fd9faa52d6a0 😄

**Comment by swlasse** [comment]
+1. This is a good feature request. Would be great to see support for non-seekable streams in the TransferUtility class.

**Comment by gotcreme** [comment]
Has anyone managed to fix this problem ?

**Comment by SebastianStehle** [comment]
It seems to be an S3 problem. Even when you upload it manually, you get a back a 501 from the server (Content-Length required). It cannot be solved in the SDK

**Comment by kerams** [comment]
@SebastianStehle, I realize this is the official SDK for Amazon's S3, but there are plenty of storage services that provide a compatible API (and can therefore be used with this SDK) and may not exhibit the same issue.

**Comment by SebastianStehle** [comment]
I have to support around 4-5 different storage providers anyway, but this is not a big issue. I just write everything to a temporary file and then copy it over. These days servers have SSD storage anway.

**Comment by kerams** [comment]
The very first 2 lines of the issue read:

> For large streams it is incorrect to:
> Buffer it entirely to file, when only transit needed. Not every system has resources for this.

Not only is it wasteful, but you're opening yourself to a potential denial of service.

**Comment by SebastianStehle** [comment]
I totally support this issue, but what I meant is that it is not a big issue for me. I only need it to restore backups.

For example even ASP.NET Core buffers large file uploads to files.

**Comment by lee-11** [comment]
I've worked around this issue to some extent by using multipart uploads. See the code here:

https://github.com/mlhpdx/s3-upload-stream

**Comment by scott-mac** [comment]
Just chiming in here to note that it is absolutely ridiculous that an enterprise SDK for the biggest cloud services provider on the planet does not support such a simple feature. Moreover, the fact that this ticket has been open for years and AWS has taken no steps to address the shortcoming just shows how little they care about the developer community that has to use their poorly conceived, poorly supported service

**Comment by eocron** [comment]
AWS is not so bad, they at least built a foundation protocol S3 which widely used by Cloud providers, but I agree - I switched to other providers because of this and saved my money doing so. 
Video-streams, encrypted data, zipped data - has no Content-Length, buffering them is a waste of money and security, so that's it.

**Comment by peterrsongg** [comment]
Closing this out as TransferUtility.UploadAsync no longer relies on Content-length and defaults to a chunked upload if the length is unknown.

This was released as part of S3 (3.7.202.0)
